### DISCUSSION TITLE
Roslyn Source Generator

### DISCUSSION THREAD

**Discussion by NN---** [dadhi/DryIoc#393] [Q&A]
How DryIoc compile time is compared to [StrongInject](https://github.com/YairHalberstadt/stronginject) which uses Source Generator ?
Should DryIoc use Source Generator instead of T4 ?

**Comment by dadhi** [comment]
@NN--- Thanks for asking.

At the moment DryIoc uses the T4 as you already know. 
That means you can move most of your registrations **without changes** to the compile-time,
in comparison with the source-generators where you need to **convert** your registrations into the attribute-based approach used by the StrongInject.

Ultimately, this is the biggest difference.  
- The T4-like approach provides more **flexibility** but you need to pay the price of **T4 quirkiness**.
- Source-generators is the **new standard** approach but with the **limited expressiveness** - you cannot write an arbitrary code instead, you have the fixed list of mechanisms to configure the compile-time generation.

Speaking of the compile-time DI in general, I think that it is not enough alone for the real-world scenarios, e.g. legacy systems, integrations with the third-party tools, plugging into the frameworks with the conforming APIs (like the MS.DI and ASP.NET Core). 

I think we need both the runtime and the compile-time capabilities and better both having the rich feature set. 
That means if you start from the compile-time DI functionality you probably will need to add the whole runtime DI/IoC container later and make them work well together. 

In the DryIoc, I am moving from the well-working runtime capabilities to the compile-time capabilities reusing as much as possible.
Yeah, I completely aware that T4 is bad in many ways, specifically in publishing as the NuGet package and in the end-user experience. That's why I am researching how to make the compile-time better in V5 where source-generators are still considered as one of the approaches, see https://github.com/dadhi/DryIoc/issues/212

I am open to any *crazy* ideas and help in this regard.
